File: guests/python/background.py
"""The background scene (tools/scenes/background.steps): the worker parks
for a CLICK, so a binding that used the app thread DEADLOCKS here."""

import sys
import threading

import kaya

app = kaya.App()

released = threading.Event()
posted = []
nested = []


def start():
    def worker():
        # Parks: work on the app thread would starve the release click.
        released.wait()
        for step in ("1", "2", "3"):
            def land(step=step):
                posted.append(step)
                status.set("".join(posted))

            app.post(land)

    threading.Thread(target=worker, name="background-worker", daemon=True).start()
    status.set("working")


def ping():
    alive.set("alive")


def release():
    released.set()


def nest():
    # A post from INSIDE a handler QUEUES for after; it never nests.
    nested.append("a")

    def land():
        nested.append("b")
        detail.set("".join(nested))

    app.post(land)
    nested.append("c")
    detail.set("".join(nested))


with app.window(title="background"):
    status = kaya.signal("idle")
    alive = kaya.signal("-")
    detail = kaya.signal("-")
    with kaya.column():
        kaya.label(bind=status).a11y_id("status")  # label#0
        kaya.label(bind=alive).a11y_id("alive")  # label#1
        # Addressed by id: an index read passes for an empty arm.
        kaya.label(bind=detail).a11y_id("nested")  # label#2
        kaya.button("start", on_click=start)  # button#0
        kaya.button("ping", on_click=ping)  # button#1
        kaya.button("release", on_click=release)  # button#2
        kaya.button("nest", on_click=nest)  # button#3


sys.exit(app.run())

File: guests/js/background.ts
// The background scene (tools/scenes/background.steps). THE PARK IS AN
// AWAIT: this worker has one thread, and blocking starves its own release.

import * as kaya from "kaya-gui";

const app = new kaya.App();

let releaseWorker!: () => void;
const released = new Promise<void>((resolve) => {
  releaseWorker = resolve;
});
const posted: string[] = [];
const nested: string[] = [];

function start(): void {
  async function worker(): Promise<void> {
    // Parks: work on the app thread would starve the release click.
    await released;
    // app.commit() ends each implicit transaction (docs/js-plan.md §4).
    for (const step of ["1", "2", "3"]) {
      posted.push(step);
      status.set(posted.join(""));
      await app.commit();
    }
  }

  void worker();
  status.set("working");
}

function ping(): void {
  alive.set("alive");
}

function release(): void {
  releaseWorker();
}

function nest(): void {
  // A post from INSIDE a handler QUEUES for after; it never nests.
  nested.push("a");

  function land(): void {
    nested.push("b");
    detail.set(nested.join(""));
  }

  app.post(land);
  nested.push("c");
  detail.set(nested.join(""));
}

let status!: kaya.Signal<string>;
let alive!: kaya.Signal<string>;
let detail!: kaya.Signal<string>;

app.window({ title: "background" }, () => {
  status = kaya.signal("idle");
  alive = kaya.signal("-");
  detail = kaya.signal("-");
  kaya.column(() => {
    kaya.label({ bind: status }).a11yId("status"); // label#0
    kaya.label({ bind: alive }).a11yId("alive"); // label#1
    // Addressed by id: an index read passes for an empty arm.
    kaya.label({ bind: detail }).a11yId("nested"); // label#2
    kaya.button("start", { onClick: start }); // button#0
    kaya.button("ping", { onClick: ping }); // button#1
    kaya.button("release", { onClick: release }); // button#2
    kaya.button("nest", { onClick: nest }); // button#3
  });
});

app.run();
